src/server/logging/log_with_metadata.js

```diff
@@ -0,0 +1,45 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { isPlainObject } from 'lodash';
+const symbol = Symbol('log message with metadata');
+
+export const logWithMetadata = {
+  isLogEvent(eventData) {
+    return Boolean(isPlainObject(eventData) && eventData[symbol]);
+  },
+
+  getLogEventData(eventData) {
+    const { message, metadata } = eventData[symbol];
+    return {
+      ...metadata,
```

Comment:
Are we ok with `eventData[symbol].message` shadowing `metadata.message`? Consumers might get confused if they're expecting their own `message` property to end up in the log entry.

We definitely don't want `metadata.message` to shadow `eventData[symbol].message` though.

Reply:
Unfortunately, given the current log infrastructure we can't have both a "message" and a "metadata.message" and have them both show up in the logs. Giving precedence to the "message" made the most sense to me, but perhaps we should consider throwing an error if they try to use "metadata.message".

Reply:
>  Giving precedence to the "message" made the most sense to me, but perhaps we should consider throwing an error if they try to use "metadata.message".

Given the current infrastructure, I agree with you.

If we were to throw, I'd only do it for dev mode. We wouldn't want the logger to prevent real work from happening in prod when `metadata` could literally be anything

Reply:
Yeah... and I hate differences between prod/dev mode...

Reply:
Understandable - let's leave it as-is then